ehrql/measures/calculate.py

```diff
@@ -81,11 +81,11 @@ def start(self):
 
     def check_timeout(self, interval):
         if interval != self.previous_interval:
-            elapsed_time = time.time() - self.start_time
+            self.elapsed_time = time.time() - self.start_time
             self.counter += 1
             self.previous_interval = interval
         if self.counter >= 12:
-            projected_time = elapsed_time / self.counter * self.num_iterations
+            projected_time = self.elapsed_time / self.counter * self.num_iterations
```

Comment:
I think this should fix it, but the alternative would be to only perform the `if self.counter >= 12` check when the interval changes. There's no point checking it every loop iteration because the values only change when we change interval.

You could either do that by indenting the whole block like:
```py
        if interval != self.previous_interval:
            self.previous_interval = interval
            self.counter += 1
            if self.counter >= 12:
                elapsed_time = time.time() - self.start_time
                projected_time = elapsed_time / self.counter * self.num_iterations
```

Or you could use a "guard clause" which is when you check a condition at the start of a function and then exit early if it's not met:
```py
        if interval == self.previous_interval:
            return
        self.previous_interval = interval
        self.counter += 1
        if self.counter >= 12:
            elapsed_time = time.time() - self.start_time
            projected_time = elapsed_time / self.counter * self.num_iterations
```

Reply:
> The reason the test isn't catching this is that it doesn't cover the case where the loop gets to 12 iterations and doesn't timeout. I think if you just change the mocked time calls to do 13 fast queries and then make the 14th slow it will expose the bug.

Currently when I go through the loops, the 12th iteration doesn't timeout (1x slow and 11x fast) so `projected_time` (57500.0) get's calculated once and is below `self.timeout` (259200.0). The next iteration the very slow query pushes the `projected_time` to 461538.5 and times out. Also just doing 13 fast queries doesnt expose the bug 😕

Reply:
Sorry, you're quite right. (I feel a bit like ChatGPT here: confidently asserting stuff and then apologising and correcting myself when it's pointed out I'm talking crap.)

The issue isn't that there aren't enough intervals. It's that there isn't more than _one row per interval_ because there aren't any groups. If you add this to the test measure definition (and change nothing else) then it hits the bug:
```py
        group_by=dict(
            sex=patients.sex,
        ),
```

I have actually checked that this is true this time – rather than just talking out of my arse.

Reply:
thanks @evansd 😬 I think we're finally close to merging and fixing the bug. I just wanted to check whether we both got this recent change right.

You suggested moving the line `elapsed_time = time.time() - self.start_time` into the second if statement, which means that we're only taking the time from the 12th query onwards, but that's ok I think?

Reply:
ok great! thanks for all the troubleshooting and suggestions :)